### PR TITLE
AP_OpticalFlow: support pixart 3901

### PR DESCRIPTION
**Pull request by tridge**
this is a new improved optical flow sensor from Pixart

### CODE REVIEW

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp [hunk 1]
```diff
@@ -29,6 +29,8 @@
 #include "AP_OpticalFlow_Pixart_SROM.h"
 #include <stdio.h>
 
+#define debug(fmt, args ...)  do {printf(fmt, ## args); } while(0)
```
Comment (OXINARF) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp]: Is this supposed to be always defined?

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp [hunk 2]
```diff
@@ -112,39 +116,53 @@ bool AP_OpticalFlow_Pixart::setup_sensor(void)
...
-    if (id != srom_id) {
-        printf("Pixart: bad SROM ID: 0x%02x\n", id);
+    debug("id1=0x%02x id2=0x%02x ~id1=0x%02x\n", id1, id2, uint8_t(~id1));
+    if (id1 == 0x3F && id2 == uint8_t(~id1)) {
+        model = PIXART_3900;
+    } else if (id1 == 0x49 && id2 == uint8_t(~id1)) {
+        model = PIXART_3901;
+    } else {
         goto failed;
+        debug("Not a recognised device\n");
```
Comment (OXINARF) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp]: This debug message is not going to run after a goto :slightly_smiling_face:

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp [hunk 3]
```diff
@@ -79,7 +83,7 @@ extern const AP_HAL::HAL& hal;
 AP_OpticalFlow_Pixart::AP_OpticalFlow_Pixart(OpticalFlow &_frontend) :
     OpticalFlow_backend(_frontend)
 {
-    _dev = std::move(hal.spi->get_device("external0m3"));
+    _dev = std::move(hal.spi->get_device("pixartflow"));
```
Comment (OXINARF) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp]: Is the PR missing something or am I? How does this work?

**Comment by lucasdemarchi** [comment]
@tridge what's this device? I couldn't find it.

**Comment by geofrancis** [comment]
i found the sensor but minimum order was 700 :(

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp (hunk 3 above)
Comment (tridge) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp]: my apologies, that's what it is on my test board, but this PR is driver only, not including the new spi devices for this board. I've fixed.

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp (hunk 1 above)
Comment (tridge) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp]: until this is in wider use, I think the debug is needed :-)

**Comment by OXINARF** [comment]
Merged, thanks!

**Comment by proficnc** [comment]
I have hardware coming soon for this.
We have added CAN to it, as SPI offboard is not great...

**Pull request merged**